core/src/main/java/org/jobrunr/server/carbonaware/CarbonAwareJobManager.java

```diff
@@ -0,0 +1,142 @@
+package org.jobrunr.server.carbonaware;
+
+import org.jobrunr.jobs.Job;
+import org.jobrunr.jobs.states.CarbonAwareAwaitingState;
+import org.jobrunr.utils.annotations.VisibleFor;
+import org.jobrunr.utils.mapper.JsonMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static java.lang.String.format;
+import static java.time.temporal.ChronoUnit.DAYS;
+import static java.time.temporal.ChronoUnit.HOURS;
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+import static org.jobrunr.utils.InstantUtils.isInstantBeforeOrEqualToOther;
+
+public class CarbonAwareJobManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CarbonAwareJobManager.class);
+    private static final int DEFAULT_REFRESH_TIME = 19;
+
+    private final int randomRefreshTimeOffset = ThreadLocalRandom.current().nextInt(0, 361) * 5;
+
+    private final CarbonAwareConfigurationReader carbonAwareConfiguration;
+    private final CarbonIntensityApiClient carbonIntensityApiClient;
+    private volatile CarbonIntensityForecast carbonIntensityForecast;
+    private volatile Instant nextRefreshTime;
+
+    public CarbonAwareJobManager(CarbonAwareConfiguration carbonAwareConfiguration, JsonMapper jsonMapper) {
+        this.carbonAwareConfiguration = new CarbonAwareConfigurationReader(carbonAwareConfiguration);
+        this.carbonIntensityApiClient = createCarbonIntensityApiClient(this.carbonAwareConfiguration, jsonMapper);
+        nextRefreshTime = Instant.now();
+        this.carbonIntensityForecast = new CarbonIntensityForecast();
+    }
+
+    public void updateCarbonIntensityForecastIfNecessary() {
+        if (isInstantBeforeOrEqualToOther(nextRefreshTime, Instant.now())) {
```

Comment:
I still wonder if this cannot be written as:
`now().plus(pollInterval).isAfter(nextRefreshTime)`?

Does the equal really matter? Or I must be missing something...

Reply:
> Does the equal really matter?

It can simply be before.

> now().plus(pollInterval).isAfter(nextRefreshTime)

I'm not a fan as it'd mean making the CarbonAwareJobManager aware of the poll interval. But we could change to 
`updateCarbonIntensityForecastIfNecessary` to `updateCarbonIntensityForecastIfRefreshTimeBefore(Instant)`